### PR TITLE
Add test for response headers

### PR DESCRIPTION
**Pull request by christian-bromann**
The WebDriver spec defines the response header in the [processing modal section](https://www.w3.org/TR/webdriver1/#processing-model) as follows:

> 3. Set the response’s header with name and value with the following values:
>
> Content-Type
>    "application/json; charset=utf-8"
> Cache-Control
>    "no-cache"

This patch adds tests for this case.

__Question__: the RFC defines header names as case insensitive. Should be lowercase header keys and values by default?

__Note:__: Geckodriver returns a wrong response value for content-type, so I created a bug here: https://bugzilla.mozilla.org/show_bug.cgi?id=1507428

### CODE REVIEW

File: webdriver/tests/new_session/headers.py
Comment (jgraham) [inline webdriver/tests/new_session/headers.py]: I think you should lowercase the recieved values here.

File: webdriver/tests/new_session/headers.py
Comment (jgraham) [inline webdriver/tests/new_session/headers.py]: This isn't really what this is testing. You could possibly even just use the `status` endpoint to test this since it doesn't require a session.

**Review by jgraham — DISMISSED** [review]

File: webdriver/tests/new_session/headers.py
Comment (christian-bromann) [reply webdriver/tests/new_session/headers.py]: oh dang, copy paste issue

File: webdriver/tests/new_session/headers.py
Comment (christian-bromann) [reply webdriver/tests/new_session/headers.py]: @jgraham see PR question: shall we by default lowercase key and value pairs since headers are case insensitive?

**Comment by christian-bromann** [comment]
@jgraham given the fact that this test should be applied to every response I moved the assertion into the `Session` class. I also converted header key/value pairs into lowercase per default.

**Comment by andreastt** [comment]
There also appears to be some test failures after one of the recent fixups:
https://travis-ci.org/web-platform-tests/wpt/builds/455470716?utm_source=github_status&utm_medium=notification

**Comment by christian-bromann** [comment]
@andreastt addressed outstanding issues

**Review by andreastt — APPROVED** [review]

**Comment by andreastt** [comment]
Retriggered some of the jobs and dismissed @jgraham’s old review.

**Comment by christian-bromann** [comment]
Azure Pipeline is failing due to
- Chrome not having `cache-control` in command response header
- FF Nightly returning `application/json; charset=utf8` in response header

continuous-integration/travis-ci/pr failing due to
- FF Nightly returning `application/json; charset=utf8` in response header

**Comment by andreastt** [comment]
@jgraham Aren’t those jobs supposed to run stability tests, like Travis?

**Comment by andreastt** [comment]
Or we could keep the checks in the client and emit a warning when the headers do not conform. I support the idea of moving the hard assertions to a separate test.

**Review by andreastt — CHANGES_REQUESTED** [review]
See comment above.

**Comment by christian-bromann** [comment]
@andreastt @jugglinmike I moved the assertions into the `assert_error` and `assert_success` assertion helpers which are only used in the webdriver test harness. What do you think?

**Review by jugglinmike — APPROVED** [review]
Good idea! Unlike the alternatives we've been considering, this may allow some responses to go untested. Those cases seem rare, so I think the simplicity of this approach wins out.

As for the timeout reported by Taskcluster: [unlike with the previous solution](https://github.com/web-platform-tests/wpt/pull/14074#issuecomment-439517580), the system is now capable of identifying the affected tests:

>     2018-11-19 12:17:10,747 - tc-run - INFO - Identifying tests affected in range 'HEAD^'...
>     2018-11-19 12:17:26,483 - tc-run - INFO - Identified 138 affected tests

That's a lot of tests, possibly enough to exceed our 2-hour limit, though the logs for Chrome are suspiciously short...

**Review by andreastt — APPROVED** [review]

**Comment by andreastt** [comment]
It looks like both the Chrome and Firefox stability tests timed out after two hours, whereas the regular test results passed. Is the timeout here too short when lots of tests are affected, @jgraham?

**Comment by whimboo** [comment]
@jgraham mind having a look at this issue? You may have missed @andreastt's question from Nov 20th. Thanks.

**Comment by andreastt** [comment]
Just to be sure, let’s retrigger the tests once more.

**Comment by whimboo** [comment]
There are still various job failures for TaskCluster, TravisCI, and Azure Pipelines.

**Comment by gsnedders** [comment]
Most of these look like all the brokenness we temporarily had on Friday… :/

**Comment by whimboo** [comment]
The newly added assertion has failures:
https://tools.taskcluster.net/groups/cZgXyDz3T0S0m1Sh4X024w/tasks/Ub-qkzjxRoCtmPoz6V_PNg/runs/1/logs/public%2Flogs%2Flive.log#L96762

> TypeError: assert_response_headers() takes exactly 2 arguments (1 given)

@christian-bromann can you please fix that? Thanks

**Comment by christian-bromann** [comment]
@whimboo done

- FF still fails because content type is `application/json; charset=utf8` instead of `application/json; charset=utf-8`
- Chrome still fails as it doesn't have a cache control header

**Comment by whimboo** [comment]
I don't know why it fails with Firefox given that we shipped the fix for `utf-8` with geckodriver 0.24.0. @jugglinmike do we still have to manually change the version number of the driver?

**Comment by jugglinmike** [comment]
[Nope](https://github.com/web-platform-tests/wpt/commit/480fd85811ed2e3449d48bc0c8b8ec09d0e20412)

**Comment by whimboo** [comment]
Oh wait. The Firefox job with the patch applied doesn't fail! What's failing is the verify job, which is killed by taskcluster after a runtime of 7200s.

**Comment by christian-bromann** [comment]
@whimboo would closing and reopening the PR help?

**Comment by whimboo** [comment]
Not sure. Running all the tests simply take that long. Someone might take a bit of time to check how long those verify tasks run in general, and maybe we have to bump up the maximum allowed time.

**Comment by gsnedders** [comment]
@whimboo I'm not sure we can while we're using the GitHub Taskcluster integration, but @jgraham would know better. There are a few cases where we end up running lots of tests, but I think the longer term plan is to do something smarter and dynamically shard them.

In either case, if you think this looks good, we should just force-merge this.

**Comment by mustjab** [comment]
@jgraham I see WPT WebDriver tests failing in Chrome on Windows and Linux runs with this change, and pass after reverting it locally. Anyone else seeing this?

```
  Γöé webdriver\tests\add_cookie\add.py:17:
  Γöé _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  Γöé webdriver\tests\support\asserts.py:55: in assert_error
  Γöé     assert_response_headers(response.headers)
  Γöé _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  Γöé
  Γöé headers = {'content-length': '3243', 'content-type': 'application/json; charset=utf-8'}
  Γöé
  Γöé     def assert_response_headers(headers):
  Γöé         """
  Γöé         Method to assert response headers for WebDriver requests
  Γöé
  Γöé         :param headers: dict with header data
  Γöé         """
  Γöé >       assert 'cache-control' in headers
  Γöé E       AssertionError: assert 'cache-control' in {'content-length': '3243', 'content-type': 'application/json; charset=utf-8'}
  Γöé
  Γöé headers    = {'content-length': '3243', 'content-type': 'application/json; charset=utf-8'}
  Γöé
  Γöö webdriver\tests\support\asserts.py:82: AssertionError
```

**Comment by andreastt** [comment]
@mustjab I think that would be expected, since chromedriver doesn’t return the mandated `Cache-Control` header.

See [_send a response_](https://w3c.github.io/webdriver/#dfn-send-a-response) in the spec.

**Comment by mustjab** [comment]
@andreastt So should we change the way we run these tests? We run them on a daily basis and at this moment most of the tests fail due to this error.

**Comment by andreastt** [comment]
Right, the assertion was added as a step to `assert_success()`, causing most—if not all!—of the wdspec tests to fail because the driver doesn’t return the mandated headers.

In order to avoid numerous workarounds, I think the best fix here is to insert a new line into https://cs.chromium.org/chromium/src/chrome/test/chromedriver/server/http_handler.cc?l=1125 and fix chromedriver. Alternatively we could back this change out temporarily to give you time to fix this.

**Comment by mustjab** [comment]
Thank you! I'll outreach to ChromeDriver team to fix this and for now, we'll run with this fix internally.

**Comment by jgraham** [comment]
I should have read the patch before merging it. I fully agree that this shouldn't be tested for every test but should be a specific set of tests or otherwise isolated from testing other features.

**Comment by christian-bromann** [comment]
@jgraham I will tweak the patch and make it so that it is handled in a single test instead of being part of all assertions.

**Comment by whimboo** [comment]
@jgraham, so we want to backout the already landed patch for a clean base? I don't think patching on top of the current patch is that great.

**Comment by andreastt** [comment]
If @mustjab is fixing chromedriver, I don’t necessarily see the need to move the assertion to separate test, nor a reason to back this out.

**Comment by Hexcles** [comment]
https://chromium-review.googlesource.com/c/chromium/src/+/1501997

**Pull request merged**
